### ISSUE TITLE
Title bar scaling interferes with DPI scaling on multiple monitors with different resolutions

### ISSUE THREAD

**Issue by nobk** [mintty/mintty#547]
On main monitor 5K(5120x2880), mintty for msys2 shows the correct window caption height and scroll bar width like other app, but when I drag it to second monitor 2.5K(2560x1440), the window caption height and scroll bar width became twice larger than other app in that monitor, but the font size inside the terminal is correct.
OS windows 10 or windows 8.1. 
mintty 2.2.3 (x86_64-pc-msys) and github mintty 2.3.5 (x86_64-pc-msys)

mon1 5k:
![5k](https://cloud.githubusercontent.com/assets/4659870/14934050/c2e26f1e-0ed6-11e6-9b36-3f3c692c71f8.PNG)

mon2 2.5k:
![2k5](https://cloud.githubusercontent.com/assets/4659870/14934051/ce903e86-0ed6-11e6-9484-a0c8a9322cb8.PNG)

**Comment by mintty** [comment]
This is a duplicate of #517; I'm leaving it open because the title describes it better.

**Comment by nobk** [comment]
i read #517 and rebuild mintty with the following changes, but nothing improve.
```c
    //HRESULT hr = pSetProcessDpiAwareness(Process_Per_Monitor_DPI_Aware);
    HRESULT hr = pSetProcessDpiAwareness(Process_System_DPI_Aware);
```

**Comment by nobk** [comment]
Only a self drawing custom window title bar can have proper size while DPI changed across the screen. This maybe a chance make the system menu non-block also.

**Comment by mintty** [comment]
Some links on this issue may be interesting:
* http://www.pcworld.com/article/2953978/displays/use-windows-10s-individual-display-scaling-to-perfect-your-multi-monitor-setup.html
<br>quote: “you may notice that some apps completely suck at the new resolution, while others rock”
* http://www.infoworld.com/article/2952506/microsoft-windows/high-resolution-displays-reveal-windows-10-blind-spot.html
<br>quote: “Microsoft has a quick and dirty fix called DPI virtualization... it also makes text blurry...”
<br>note: That's likely what patch #470 dealt with; I don't even know if it could be offered as an option to accept bluriness rather than title bar mis-scaling as a user-preferred trade-off because the patch involves removing some DPI magic from the compiled-in “manifest”.
* https://www.reddit.com/r/Surface/comments/2uzjp6/ok_im_seriously_done_with_these_scaling_issues/
<br>quote: “I'm seriously done with these scaling issues”
<br>note: And so am I. I am not going to investigate more research into MS's API to find a workaround for their crappy DPI handling.

Thanks for your hint that a self-made title bar would be needed for a fix, that's what I suspected already. If anyone wants to provide an implementation, you are welcome.

**Comment by HowellBP** [comment]
I found this: http://www.danantonielli.com/adobe-app-scaling-on-high-dpi-displays-fix/

Apparently adding a manifest file for mintty.exe and setting the registry flag HKEY_LOCAL_MACHINE > SOFTWARE > Microsoft > Windows > CurrentVersion > SideBySide > DWORD:PreferExternalManifest:1 fixes the issue. I wrote a manifest which works properly and makes the window wrapper scale-aware for the app: https://gist.github.com/HowellBP/4caa1995b296c45e0739bd91529aa5da

Putting that in /bin/mintty.exe.manifest along with setting the registry flag fixes the DPI awareness. This can be a temporary workaround for anyone else having this issue until the root cause is fixed. Hope this helps.

**Comment by HowellBP** [comment]
It looks like you could get away with adding
```
<windowsSettings>
    <ms_windowsSettings:dpiAware xmlns:ms_windowsSettings="http://schemas.microsoft.com/SMI/2005/WindowsSettings">false</ms_windowsSettings:dpiAware>
</windowsSettings>
```
and to src/res.mft and that SHOULD solve this problem. I'll try compiling myself once I get an environment set up to do so.

**Comment by mintty** [comment]
Note issue #470; a DPI-related entry was removed from that same file to achieve unblurred rescaling of the text. The proposed section looks differently, and especially sets some dpiAware value to false rather than true as it used to be, so maybe that's a solution, but we need to check against the blurriness problem.
Why is this stuff so crappy?

**Comment by HowellBP** [comment]
In my initial tests with just the manifest file (Windows 10 x64) on a 1920x1200 external monitor, I didn't see any noticeable blurriness with the text, even scaling it up and down in the terminal. It might be more related to ClearText antialiasing, but that's just a guess. And I agree, it is crappy.

I tested compiling it with the altered res.mft and it started up with no problems, but I won't be able to test on another display until Monday.

**Comment by HowellBP** [comment]
I just confirmed that the modified section of the compiled-in res.mft works, at least on Windows 10 x64. Can anyone else with this issue compile and test with the above modification?

**Comment by mintty** [comment]
@espresso3389: Can you check/comment this modification, please? (I can only simulate a test case, using a Full-HD monitor scaled down.)

**Comment by HowellBP** [comment]
Here's the section of res.mft that needs to be modified:
```
      <supportedOS Id="{8e0f7a12-bfb3-4fe8-b9a5-48fd50a15a9a}"/> <!-- 10 -->
    </application>
  </compatibility>
  <application>
    <windowsSettings>
      <ms_windowsSettings:dpiAware xmlns:ms_windowsSettings="http://schemas.microsoft.com/SMI/2005/WindowsSettings">false</ms_windowsSettings:dpiAware>
    </windowsSettings>
  </application>
</assembly>
```

**Comment by nobk** [comment]
I just apply HowellBP's patch, its working well on my environment. Drag new compile mintty across two screens, the title bar size is correct.

**Comment by nobk** [comment]
And the bad news is, after the patch , the font rendering quality  become lower than before(the blurriness problem), that is unacceptable.

**Comment by HowellBP** [comment]
@nobk I don't understand this font rendering issue you're having. I have a 3840x2160 display on my laptop and a 1920x1200 external, and the fonts look exactly the same dragging the window between them with my patch applied. Can you give some screenshots? What font configuration do you have (font type, size)? What are your font DPI and ClearType settings? Is "Font Smoothing" set to something besides "Full" in your MinTTY options?

**Comment by nobk** [comment]
@HowellBP here is the screenshot, on 5120x2880 screen.
![z](https://cloud.githubusercontent.com/assets/4659870/15126508/e9bbf638-1664-11e6-8ea5-f6ee63fbf508.PNG)

**Comment by HowellBP** [comment]
Okay, so it looks like the fix I found causes DPI awareness to be ignored completely, which apparently includes font rendering. I guess this is something that will have to be handled by an API call rather than the manifest.

**Comment by mintty** [comment]
Try winmain.c, line 1997:
```
  per_monitor_dpi_aware = set_per_monitor_dpi_aware();
```
→
```
//  per_monitor_dpi_aware = set_per_monitor_dpi_aware();
  per_monitor_dpi_aware = true;
```
with the res.mft patch; does that help?

**Comment by nobk** [comment]
@mintty To avoid warning -Werror=unused-function, i add #if 0 to function set_per_monitor_dpi_aware() before line 158 , compile succeed. But still the same font rendering problem inside the terminal window.
```
//  per_monitor_dpi_aware = set_per_monitor_dpi_aware();
  per_monitor_dpi_aware = true;
```

**Comment by HowellBP** [comment]
I think I got it, and I feel kind of dumb for not trying this sooner.
![image](https://cloud.githubusercontent.com/assets/495667/15129602/2f2495de-1608-11e6-80ed-5f5935690917.png)

I changed the dpiAware flag in res.mft to "true". It now works perfectly switching resolutions between high-DPI and low-DPI displays regardless of Windows' "scale". So your `set_per_monitor_dpi_aware()` function works correctly, I just didn't try flipping the dpiAware bit in the manifest sooner.

**Comment by nobk** [comment]
good job @HowellBP.

**Comment by HowellBP** [comment]
The new res.mft diff:
```
diff --git a/src/res.mft b/src/res.mft
index 509babd..daffa63 100644
--- a/src/res.mft
+++ b/src/res.mft
@@ -30,4 +30,9 @@
       <supportedOS Id="{8e0f7a12-bfb3-4fe8-b9a5-48fd50a15a9a}"/> <!-- 10 -->
     </application>
   </compatibility>
+  <application>
+    <windowsSettings>
+      <ms_windowsSettings:dpiAware xmlns:ms_windowsSettings="http://schemas.microsoft.com/SMI/2005/WindowsSettings">true</ms_windowsSettings:dpiAware>
+    </windowsSettings>
+  </application>
 </assembly>
```

**Comment by mintty** [comment]
I cannot test with monitors of significantly different resolution, but I tried with two full-HD monitors, one of them scaled up to the maximum of 175%, and I am seeing the bluriness problem with either way of the patch (false or true).

**Comment by mintty** [comment]
The problem is, as soon as *anything* about DPI is set in the manifest,
we are not getting the WM_DPICHANGED message anymore,
see also https://github.com/mintty/mintty/pull/470#issuecomment-138293776.
If anyone can find out why or rather how it can be enforced that would be great!

**Comment by mintty** [comment]
Except (http://stackoverflow.com/questions/32115963/not-getting-the-wm-dpichanged-notification) if we set the value to true/PM but then the title bar isn't scaled anymore!$§@!

The problem is also discussed in http://stackoverflow.com/questions/33350881/auto-scale-but-still-process-wm-dpichanged and they even seem to have a workaround! It's actual implementation however looks light a nightmare and I don't feel like deriving a patch for mintty from it.

**Comment by mintty** [comment]
There is also a discussion in https://emoacht.wordpress.com/2013/10/ but there it seems as if a manifest setting and explicit WM_DPICHANGED handling could coexist, which doesn't appear to be working in mintty.

**Comment by mintty** [comment]
Some new observations (based on my test setup, see above):

1. We need this in res.mft:
```
  <application>
    <windowsSettings>
      <dpiAware xmlns="http://schemas.microsoft.com/SMI/2005/WindowsSettings">false</dpiAware>
    </windowsSettings>
  </application>
```
to activate Windows’ auto-scaling of the title bar. But then, due to Windows’ insane design, WM_DPICHANGED is not sent anymore, so the auto-scaled font stays blurred.
However, WM_CAPTURECHANGED is sent when moving a window to another screen and this causes win_adapt_term_size to be  called.

2. This isn’t actually a problem of moving a window to another screen. If mintty is started on the DPI-scaled screen right away, the font is blurred, too.

So we don’t actually need to change the message handling. Rather DPI should be considered properly in win_adapt_term_size (hopefully using some of the information from the link collection above).

**Comment by AlexSamad** [comment]
Hi 

having the same issue. 

Only thing I would like to point out is that google chrome handles the high dpi / high res screens well.

mintty and also windows cmd  don't

Also the cygwin rxvt-native used to handle it appropiately.

**Comment by AlexSamad** [comment]
Interestingly i found this. I was installing a git client for windows.

The picutre shows the 2 versions of mintty the MINGW64 is able to handle the different frame sizes.
The newer mintty version doesn't

Notice the fonts sizes are correct in 2.0.3 and not in 2.3.7

![image](https://cloud.githubusercontent.com/assets/5509219/15560204/a2a522d6-232c-11e6-87b4-981bcd266153.png)

**Comment by mintty** [comment]
Read the previous comments in this issue, especially
```
“I'm seriously done with these scaling issues”
```
and you will see that it does not help to point to other software that may handle this (especially if it has lots of manpower behind and handles the title bar itself anyway); it would help if you make an implementation proposal.

DPI handling was introduced since 2.0.3 to ensure the scaled font has clear, not blurred display. It also was discussed before in this issue that the insane Windows API design makes it hard if not impossible to have proper title bar scaling and sharp font display at the same time.

About font size scaling, I think most people are satisfied with the handling after recent changes. If you think it's not correct, please define what you mean and what "correct" handling is.

**Comment by nobk** [comment]
On my environment this bug has been fixed, with  HowellBP's patch 2.
I tested on both cygwin and msys2, working fine.

**Comment by mintty** [comment]
According to my own testing, that patch fixes the title bar problem but recovers the blurred font problem, so that would not be a solution.
May the different test scenario obscures the issue? Maybe the bluriness is not easily visible on a 4K monitor just *because* of the high resolution? (Maybe someone would buy me a 4K monitor to reproduce the test case...?)
Whatever, I cannot integrate a patch that exhibits bad behaviour here, sorry.

**Comment by nobk** [comment]
Yes maybe we are not using the same environment, I am using the 3rd part MacType as windows font renderer, not windows clear type, maybe windows clear type always blurriness when show any character. My screen shot on  5120x2880 screen.
![1](https://cloud.githubusercontent.com/assets/4659870/15582163/1ff88546-23a2-11e6-8f56-db64f8d9e7a9.PNG)

And on 2568x1440 screen.
![2](https://cloud.githubusercontent.com/assets/4659870/15582584/4c58c7e8-23a4-11e6-8a4f-a0ec2d8ac0c3.png)

**Comment by nobk** [comment]
git for windows also using mintty as the 'Git bash here' right mouse menu on file explorer.
and I find the version they using, or they build, is not have this issue. But I can not build that version with my msys2.
https://github.com/git-for-windows/MSYS2-packages/blob/master/mintty/PKGBUILD

`pkgver=2.3.3`
`...`
`source=(${pkgname}-${pkgver}.tar.gz::https://github.com/mintty/mintty/archive/${pkgver}.tar.gz)`

**Comment by mintty** [comment]
Git for Windows uses an outdated release of mintty. It does not yet have the DPI fix which means it has the font blurring problem which is worse than the title bar problem.

**Comment by AlexSamad** [comment]
Add my 2c

This is git-bash window side by side with the new mintty version
Notice the larger text thinks its 9pnt and he smaller (git-bash) is 12 point

[image: Inline images 1]

**Comment by mintty** [comment]
After some more testing of configuration combinations, it is pretty clear what had been observed step-by-step already:
- The title bar is only scaled properly with a manifest entry dpiAware, value `false` or `true`.
- With these two values, scaled fonts look blurred (unless your dpi is so high that you don't see it...).

There are only two solutions:
- Nag Microsoft to fix this incredibly stupid bug.
- Draw the title bar (and scrollbar) by mintty rather than Windows.

By the way, the issue can be masked out by running mintty without title bar:
`mintty -B frame` or `mintty -B void`.

**Comment by AlexSamad** [comment]
Can you not add a command line option to allow the user to decide.

![image](https://cloud.githubusercontent.com/assets/5509219/15729487/01c437ae-28a8-11e6-81ff-3e214d8280f9.png)


Back window - normal windows started on my laptop (where it sort of looks okay) dragged to external window
front left is normal windows dragged with shift to external window
front right it git-bash window - its the only one that is actually working well.

they are all 9point lucinda console

I like the git -bash windows and the way it works it seems to be fine ..

**Comment by mintty** [comment]
Unfortunately, proper title bar scaling depends on that manifest entry, so it's statically configured.
If someone tells me how that could be switched dynamically, there could a an option for it.
Reportedly, the manifest can be added with a tool, or even as a separate file. I don't know whether that works if there is a compiled-in manifest already, and how it would be done. If someone points that out, there could perhaps also be a wrapper script.

**Comment by mintty** [comment]
About the font size visually larger than "9pt", this is actually a different issue, discussed in #492.

**Comment by mintty** [comment]
Actually, I can make the whole DPI-dependent adaptation handling optional.
This should affect both title bar and font size; whether it's proper handling, on the other hand, is questionable, see https://github.com/mintty/mintty/issues/492#issuecomment-223807119.

**Comment by mintty** [comment]
Option to ignore DPI changes release in 2.4.0.
Proper behaviour of DPI handling explained above otherwise (with no further comments), so closing.

**Comment by mintty** [comment]
Another link I found describes the issue well, including the interference of blurriness and title bar scaling:
http://stackoverflow.com/questions/36864894/scaling-the-non-client-area-title-bar-menu-bar-for-per-monitor-high-dpi-suppo/37624363

**Comment by nobk** [comment]
@mintty To avoid recompile mintty with manifeset patch, i create a mitty.exe.manifest xml file with <ms_windowsSettings:dpiAware xmlns:ms_windowsSettings="http://schemas.microsoft.com/SMI/2005/WindowsSettings">true</ms_windowsSettings:dpiAware>.
and put it to /usr/bin/ where put mintty.exe, turn on registry switch to enable external manifest file for exe.
C:\Windows\system32>reg add HKLM\Software\Microsoft\Windows\CurrentVersion\SideBySide /v PerferExternalManifest /t REG_DWORD /d 1
The last version of mintty for msys2 work well without blurriness on 5k screen and no border size issue on 2.5k screen, with the manifest xml which set dpiAware to true.

**Comment by mintty** [comment]
The manifest options have been discussed above already; unfortunately the same values that result in proper title bar scaling (false or true, strangely enough) result in pixel-based scaling of the rendered font display, thus in blurred text display (as you noted yourself in https://github.com/mintty/mintty/issues/547#issuecomment-217970542).
This is described and demonstrated in the article I referenced in my previous comment. The blurriness may in fact be compensated by the high resolution, so you may not see it on those monitors, but yet it's there, and some people don't like it, so that's not an option for mintty.

It had previously been claimed that a dpiAware value of "true" would fix the problem (https://github.com/mintty/mintty/issues/547#issuecomment-218006767 and your later response) but that does not happen on my Windows 10 test system. Maybe it's different with Windows 8?

The good news here is that MS has finally recognised the annoyance factor of their immature DPI handling model (https://blogs.technet.microsoft.com/askcore/2016/08/16/display-scaling-changes-for-the-windows-10-anniversary-update/#comment-60245 and new function EnableNonClientDpiScaling in https://msdn.microsoft.com/en-us/library/windows/desktop/hh447398%28v=vs.85%29.aspx?f=255&MSPPError=-2147217396). Unfortunately, for obscure reasons, they prefer not to enhance the existing handling but rather introduced an overly complex additional API with obscure prerequisites and it's not even fully documented yet.

Anyway, as I understand, this new MS function may be suited to finally resolve the issue. If someone wants to have this tracked here, I'd prefer to open a new issue rather than reopen this one.

**Comment by nobk** [comment]
I did pixel level compare on my windows 10/8.1 screen 5k, no blurriness issue when set dpiAware to true. Maybe the different environment we using, I am using Mactype as the default font renderer, not the ms windows build in cleartype.

**Comment by mintty** [comment]
Ah, yes, you had mentioned this detail before (see above), I forgot. This is actually interesting, someone should point that out to MS and ask them to implement proper font rendering for auto-scaled windows.

For mintty however, this does not help, as I cannot tune it for a non-standard environment at the cost of making things worse in a standard configuration.

**Comment by mintty** [comment]
Released revised DPI handling in 2.5.1; EnableNonClientDpiScaling is being used as documented but it does not work.

**Comment by mintty** [comment]
This issue is finally resolved, #588.

**Comment by mintty** [comment]
Released 2.6.1.

**Comment by mintty** [comment]
Windows 10 build 17115 introduces a new setting below the scaling factor, Advanced scaling settings, to "fix" scaling for apps. If enabled, this should be fixing some remaining scaling issues, at least for new windows, with title bar, menus and Options dialog.

**Comment by mintty** [comment]
With support for Windows DPI handling V2, I hope all these problems are now resolved.

**Comment by mintty** [comment]
Released 2.9.0.
